File: slack-reader/src/main/java/dev/number6/slackreader/SlackService.java
package dev.number6.slackreader;

import dev.number6.slackreader.model.Message;
import dev.number6.slackreader.model.WorkspaceMessages;
import dev.number6.slackreader.port.SlackPort;
import dev.number6.slackreader.port.SlackReaderConfigurationPort;
import com.amazonaws.services.lambda.runtime.LambdaLogger;

import java.time.LocalDate;
import java.util.Collection;
import java.util.stream.Collectors;

public class SlackService {
    private final SlackPort slackPort;
    private final SlackReaderConfigurationPort config;

    public SlackService(SlackPort slackPort, SlackReaderConfigurationPort config) {

        this.slackPort = slackPort;
        this.config = config;
    }

    public WorkspaceMessages getMessagesOnDate(LocalDate comprehensionDate, LambdaLogger logger) {
        Collection<String> blacklistedChannels = config.getBlacklistedChannels();

        return slackPort.getChannelList(logger).stream()
                .filter(c -> !blacklistedChannels.contains(c.getName()))
                .reduce(new WorkspaceMessages(comprehensionDate),
                        (wm, c) -> {
                            wm.add(c.getName(),
                                    slackPort.getMessagesForChannelOnDate(c, comprehensionDate, logger)
                                            .stream()
                                            .map(Message::getText)
                                            .collect(Collectors.toList()));
                            return wm;
                        },
                        (wm1, wm2) -> wm2);
    }
}

File: slack-reader/src/main/kotlin/dev/number6/slackreader/SlackService.kt
package dev.number6.slackreader

import com.amazonaws.services.lambda.runtime.LambdaLogger
import dev.number6.slackreader.model.WorkspaceMessages
import dev.number6.slackreader.port.SlackReaderPort
import dev.number6.slackreader.port.SlackReaderConfigurationPort
import java.time.LocalDate

class SlackService(private val slackReaderPort: SlackReaderPort, private val config: SlackReaderConfigurationPort) {
    fun getMessagesOnDate(comprehensionDate: LocalDate, logger: LambdaLogger): WorkspaceMessages {
        val blacklistedChannels = config.getBlacklistedChannels()
        return slackReaderPort.getChannelList(logger)
                .filter { c -> !blacklistedChannels.contains(c.name) }
                .fold(WorkspaceMessages(comprehensionDate),
                        { wm, c ->
                            wm.add(c.name,
                                    slackReaderPort.getMessagesForChannelOnDate(c, comprehensionDate, logger)
                                            .map { m -> m.text })
                        })
    }
}